24.0.0.9 staging (#4022)
OpenLiberty/blogs: posts/2024-09-10-24.0.0.9.adoc

apply plugin: 'liberty'
----

Or if you're using link:{url-prefix}/docs/latest/container-images.html[container images]:

[source]
----
FROM icr.io/appcafe/open-liberty
----

Or take a look at our link:{url-prefix}/start/[Downloads page].

If you're using link:https://plugins.jetbrains.com/plugin/14856-liberty-tools[IntelliJ IDEA], link:https://marketplace.visualstudio.com/items?itemName=Open-Liberty.liberty-dev-vscode-ext[Visual Studio Code] or link:https://marketplace.eclipse.org/content/liberty-tools[Eclipse IDE], you can also take advantage of our open source link:https://openliberty.io/docs/latest/develop-liberty-tools.html[Liberty developer tools] to enable effective development, testing, debugging, and application management all from within your IDE.

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/29558
// Contact/Reviewer: yasmin-aumeeruddy
// // // // // // // //
[#mptelem]
== Manage your logs, metrics, and traces with MicroProfile Telemetry 2.0

> In this release, the link:{url-prefix}/docs/latest/reference/feature/mpTelemetry-2.0.html[MicroProfile Telemetry 2.0] feature (`mpTelemetry-2.0`) helps improve the observability of your Java applications by using OpenTelemetry to collect and export logs, metrics, and traces in a standardized way. Previous versions of Micro Profile Telemetry could manage only distributed tracing.

Reviewer: ```suggestion
In this release, the link:{url-prefix}/docs/latest/reference/feature/mpTelemetry-2.0.html[MicroProfile Telemetry 2.0] feature (`mpTelemetry-2.0`) helps improve the observability of your Java applications by using OpenTelemetry to collect and export logs, metrics, and traces in a standardized way. Previous versions of MicroProfile Telemetry could manage only distributed tracing.
```

For more information about the available configuration properties, see link:{url-prefix}/docs/latest/microprofile-config-properties.html#telemetry[MicroProfile Config properties: MicroProfile Telemetry].

For more information about using MicroProfile Telemetry to manage your metrics, logs, and traces in a standardized way, see link:{url-prefix}/docs/latest/microprofile-telemetry.html[Enable observability with MicroProfile Telemetry].

> For information about managing <<logs, logs>> and <<metrics, metrics>> with MicroProfile Telemetry, see the following sections.

Reviewer: ```suggestion
For information about managing <<metrics, metrics>> and <<logs, logs>> with MicroProfile Telemetry, see the following sections.
```

This lines says `logs` and `metrics`, but then the order of them in the following section are `metrics` and `logs`.  Shouldn't it be consistent?

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/29563
// Contact/Reviewer: Channyboy
// // // // // // // //
[#metrics]
=== Send Liberty metrics to OpenTelemetry

> MicroProfile Telemetry 2.0 provides the latest OpenTelemetry technology. In addition to link:{url-prefix}/docs/latest/microprofile-telemetry.html[distributed tracing], the feature now allows OpenTelemetry to collect and export metrics and logs.

Reviewer: I almost want to have this sentence to be before the last sentence and moved above, i.e.

```
MicroProfile Telemetry 2.0 provides the latest OpenTelemetry technology. In addition to link:{url-prefix}/docs/latest/microprofile-telemetry.html[distributed tracing], the feature now allows OpenTelemetry to collect and export metrics and logs.   For information about managing <<metrics, metrics>> and <<logs, logs>> with MicroProfile Telemetry, see the following sections.
```

> When the MicroProfile Telemetry 2.0 feature (`mpTelemetry-2.0`) is enabled, Open Liberty can now forward runtime component statistics that are captured by the link:{url-prefix}/docs/latest/reference/feature/monitor-1.0.html[Performance Monitoring 1.0] feature (`monitor-1.0`) to the MicroProfile Telemetry 2.0 runtime. This statistical data is registered as metrics in the telemetry runtime and can then be forwarded to any OpenTelemetry Protocol (OTLP) compatible metric consumer to meet your monitoring needs.

Reviewer: ```suggestion
When the MicroProfile Telemetry 2.0 feature (`mpTelemetry-2.0`) is enabled, Open Liberty can now forward runtime component statistics that are captured by the link:{url-prefix}/docs/latest/reference/feature/monitor-1.0.html[Performance Monitoring 1.0] feature (`monitor-1.0`) to the MicroProfile Telemetry 2.0 runtime. These statistical data are registered as metrics in the telemetry runtime and can be forwarded to any OpenTelemetry Protocol (OTLP) compatible metric consumer to meet your monitoring needs.
```

I know that `This data is` can be "correct" at times, so you can leave it as well.  The `then` though really is not needed.

For example, the `ConnectionPool` component requires the following configuration:

[source,xml]
----
<featureManager>
   <feature>mpTelemetry-2.0</feature>
   <feature>jdbc-4.3</feature>
</featureManager>
----

By default, all OpenTelemetry data is exported to link:https://opentelemetry.io/docs/languages/java/exporters/#otlp[OTLP]. You can set a different exporter by specifying the following system property or environment variable:

* System property: `otel.metrics.exporter`
* Environment variable: `OTEL_METRICS_EXPORTER`

You can also optionally configure the metric export interval configuration variable. The value is specified in milliseconds and the default is 60000 (60 seconds).

* System property: `otel.metric.export.interval`
* Environment variable: `OTEL_METRIC_EXPORT_INTERVAL`

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/29551
// Contact/Reviewer: pgunapal
// // // // // // // //
[#logs]
=== Send logs to OpenTelemetry

> The `mpTelemetry-2.0` feature can now collect Open Liberty runtime log sources (messages, traces, ffdcs) and application logs generated through the `java.util.logging` package (JUL).

Reviewer: ```suggestion
The `mpTelemetry-2.0` feature can now collect Open Liberty runtime log sources (messages, traces, ffdcs) and application logs generated through the `java.util.logging` (JUL) package.
```
Maybe it was ok the way it was, but it seems appropriate to put the abbreviation after the package name.

To enable the MicroProfile Telemetry 2.0 feature to collect all logs, add the following configuration to your `server.xml` file:

[source,xml]
----

> <features>
>    <feature>mpTelemetry-2.0</feature>
> </features>

Reviewer: ```suggestion
<featureManager>
   <feature>mpTelemetry-2.0</feature>
</featureManager>
```

----

If the `mpTelemetry` configuration element or the `source` attribute is not configured, the `message` source is set by default. In this case, only messages are collected. If the `source` attribute is specified empty (`source=""`), no logs are sent to OpenTelemetry.

To collect and export runtime-level logs, enable OpenTelemetry by using the following system property or environment variable:

* System property: `otel.sdk.disabled=false`
* Environment variable: `OTEL_SDK_DISABLED=false`

You can set configuration properties in any of the link:{url-prefix}/docs/latest/external-configuration.html#default[config sources that are available to MicroProfile Config].

To separately configure multiple applications in a server, you can configure OpenTelemetry with application configuration. However, you cannot collect runtime-level logs this way.

By default, all OpenTelemetry data is exported to link:https://opentelemetry.io/docs/languages/java/exporters/#otlp[OTLP]. You can set a different exporter by specifying the following system property or environment variable:

* System property: `otel.logs.exporter`
* Environment variable: `OTEL_LOGS_EXPORTER`

> For more information about the available configuration properties, see xref:{url-prefix}/docs/latest/microprofile-config-properties.html#telemetry[MicroProfile Config properties: MicroProfile Telemetry].

Reviewer: Should this sentence be at the end of the metrics section as well?

   <samesite none="*" partitioned="true|false"/>
</httpEndpoint>
----

Alternatively, you can set `Partitioned` by using the `Set-Cookie` header with the following two `HttpServletResponse` APIs:

* link:https://openliberty.io/docs/latest/reference/javadoc/liberty-jakartaee10-javadoc.html?path=liberty-jakartaee10-javadoc/jakarta/servlet/http/HttpServletResponse.html[HttpServletResponse.setHeader]
* link:https://openliberty.io/docs/latest/reference/javadoc/liberty-jakartaee10-javadoc.html?path=liberty-jakartaee10-javadoc/jakarta/servlet/http/HttpServletResponse.html[HttpServletResponse.addHeader]

For more information, including a visual example, see link:https://github.com/privacycg/CHIPS?tab=readme-ov-file#chips-cookies-having-independent-partitioned-state[CHIPS (Cookies Having Independent Partitioned State)] on GitHub.

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/29571
// Contact/Reviewer: jhanders34
// // // // // // // //
[#versionless]
== Versionless features for Java / Jakarta EE Container Liberty features

> In 24.0.0.8, Open Liberty introduced versionless Java EE and Jakarta EE features.  Those new versionless features allow you to easily use features without needing to know what feature version to use.

Reviewer: Didn't know if you wanted to add a reference link back to https://openliberty.io/blog/2024/08/13/24.0.0.8.html#versionless

Reviewer 2: I linked to the versionless features doc at the end of this section instead

> In 24.0.0.8, Open Liberty introduced versionless Java EE and Jakarta EE features.  Those new versionless features allow you to easily use features without needing to know what feature version to use.
> 
> In 24.0.0.8, the versionless features did not include the `Container` features, which enable you to provide your own implementation of that Java EE or Jakarta EE component specification.  An example of such a feature is `facesContainer-4.0`.

Reviewer: ```suggestion
In 24.0.0.8, Open Liberty introduced versionless Java EE and Jakarta EE features.  Those new versionless features allow you to easily use features without needing to know what feature version to use.  The initial release of versionless features did not include the `Container` features, which enable you to provide your own implementation of that Java EE or Jakarta EE component specification.  An example of such a feature is `facesContainer-4.0`.
```

In 24.0.0.9, Open Liberty adds versionless features for the missing `Container` features. The following versionless features are added:

- `jpaContainer` / `persistenceContainer`
- `jsfContainer` / `facesContainer`
- `jsonbContainer`

> - `jspContainer`

Reviewer: ```suggestion
- `jsonpContainer`
```
